src/api/workspace.rs

```diff
@@ -1960,18 +1961,69 @@ async fn post_database_row_handler(
     .enforce_action(&uid, &workspace_id, Action::Write)
     .await?;
 
+  let AddDatatabaseRow { cells, document } = add_database_row.into_inner();
+
   let new_db_row_id = biz::collab::ops::insert_database_row(
     &state.collab_access_control_storage,
     &state.pg_pool,
     &workspace_id,
     &db_id,
     uid,
-    cells_by_id.into_inner(),
+    None,
+    cells,
+    document,
   )
   .await?;
   Ok(Json(AppResponse::Ok().with_data(new_db_row_id)))
 }
 
+async fn put_database_row_handler(
+  user_uuid: UserUuid,
+  path_param: web::Path<(String, String)>,
+  state: Data<AppState>,
+  upsert_db_row: Json<UpsertDatatabaseRow>,
+) -> Result<Json<AppResponse<String>>> {
+  let (workspace_id, db_id) = path_param.into_inner();
+  let uid = state.user_cache.get_user_uid(&user_uuid).await?;
+  state
+    .workspace_access_control
+    .enforce_action(&uid, &workspace_id, Action::Write)
+    .await?;
+
+  let UpsertDatatabaseRow {
+    pre_hash,
+    cells,
+    document,
+  } = upsert_db_row.into_inner();
+
+  let row_id = {
+    let mut hasher = Sha256::new();
+    hasher.update(workspace_id.clone());
```

Comment:
IIRC, the update function is `fn update(&mut self, data: impl AsRef<[u8]>)`
So you can remove clone

Reply:
done